File: examples/go-clean-arch/adapter/stripe/payment.go
package stripe

import (
	"fmt"

	"github.com/example/order-service/domain"
)

// Payment implements port.PaymentGateway.
// Corresponds to: adapter StripePayment : framework implements PaymentGateway
//   inject client: *stripe.Client
//
// Stub implementation for concept verification.
type Payment struct {
	nextTxID int
}

func NewPayment() *Payment {
	return &Payment{nextTxID: 1}
}

func (p *Payment) Charge(amount domain.Money, token string) (string, error) {
	if token == "" {
		return "", fmt.Errorf("payment token required")
	}
	if amount.Amount <= 0 {
		return "", fmt.Errorf("charge amount must be positive")
	}
	txID := fmt.Sprintf("tx_%d", p.nextTxID)
	p.nextTxID++
	return txID, nil
}

func (p *Payment) Refund(txID string) error {
	if txID == "" {
		return fmt.Errorf("transaction ID required")
	}
	return nil
}

File: examples/go-clean-arch/Arch/Payment.hs
module Arch.Payment (paymentRecord, paymentGateway, stripePayment, declareAll) where

import Plat.Core
import Plat.Ext.CleanArch

import Arch.Shared (money)

paymentRecord :: Decl 'Model
paymentRecord = model "PaymentRecord" enterprise $ do
  field "id"            (customType "UUID")
  field "orderId"       (customType "UUID")
  field "amount"        (ref money)
  field "method"        string
  field "status"        string
  field "transactionId" string

paymentGateway :: Decl 'Boundary
paymentGateway = port "PaymentGateway" interface $ do
  op "charge" ["amount" .: ref money, "token" .: string] ["txId" .: string, "err" .: error_]
  op "refund" ["txId" .: string] ["err" .: error_]

stripePayment :: Decl 'Adapter
stripePayment = impl_ "StripePayment" framework paymentGateway $ do
  inject "client" (ext "*stripe.Client")

declareAll :: ArchBuilder ()
declareAll = do
  declare paymentRecord
  declare paymentGateway
  declare stripePayment
